presto-main/src/main/java/com/facebook/presto/operator/StreamingAggregationOperator.java

```diff
@@ -0,0 +1,294 @@
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.systemMemoryContext = operatorContext.newLocalSystemMemoryContext();
+        this.userMemoryContext = operatorContext.localUserMemoryContext();
+        this.groupByTypes = ImmutableList.copyOf(requireNonNull(groupByTypes, "groupByChannels is null"));
+        this.groupByChannels = ImmutableList.copyOf(requireNonNull(groupByChannels, "groupByChannels is null"));
+        this.accumulatorFactories = requireNonNull(accumulatorFactories, "accumulatorFactories is null");
+        this.step = requireNonNull(step, "step is null");
+
+        this.aggregates = setupAggregates(step, accumulatorFactories);
+        this.pageBuilder = new PageBuilder(toTypes(groupByTypes, aggregates));
+    }
+
+    private List<Aggregator> setupAggregates(Step step, List<AccumulatorFactory> accumulatorFactories)
+    {
+        ImmutableList.Builder<Aggregator> builder = ImmutableList.builder();
+        for (AccumulatorFactory factory : accumulatorFactories) {
+            builder.add(new Aggregator(factory, step));
+        }
+        return builder.build();
+    }
+
+    private static List<Type> toTypes(List<Type> groupByTypes, List<Aggregator> aggregates)
+    {
+        ImmutableList.Builder<Type> builder = ImmutableList.builder();
+        builder.addAll(groupByTypes);
+        aggregates.stream().forEach(aggregator -> builder.add(aggregator.getType()));
+        return builder.build();
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing && pendingInput == null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(pendingInput == null);
+        requireNonNull(page, "page is null");
+
+        processInput(page);
+        updateMemoryUsage();
+    }
+
+    private void updateMemoryUsage()
+    {
+        long memorySize = pageBuilder.getRetainedSizeInBytes();
+        for (Aggregator aggregator : aggregates) {
+            memorySize += aggregator.getEstimatedSize();
+        }
+
+        if (previousRow != null) {
+            memorySize += previousRow.getRetainedSizeInBytes();
+        }
+
+        if (step.isOutputPartial()) {
+            systemMemoryContext.setBytes(memorySize);
+        }
+        else {
+            userMemoryContext.setBytes(memorySize);
+        }
+    }
+
+    private void processInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+
+        if (previousRow != null) {
+            if (!rowEqualsRow(0, previousRow, 0, page)) {
+                evaluateAggregatesToPageBuilder(previousRow, 0);
+                previousRow = null;
+
+                if (pageBuilder.isFull()) {
+                    pendingInput = page;
```

Comment:
My idea was to not have any `pendingInputs` at all (but some `outputPages` list) here as it makes `getOutput` logic tricky